Feat: Docusaurus powered docs served to gh-pages

Xiltyn opened the pull request:
### Summary

1. Added Docusaurus v2.1 to the project inside the `./docusaurus` directory
2. Moved readMe content into the new docs
3. Added test gh-actions for deploying docs into `https://callstack.github.io/reassure/` gh-page
4. Fixed eslint related issues on the main pipeline (node ver bump to 16.x | yarn install step for ./docusaurus directory)

mdjastrzebski commented on .github/workflows/docs-deploy-test.yml:
```diff
@@ -0,0 +1,24 @@
+on:
+  pull_request:
+    branches:
+      - feature/docusaurus-based-docs
```
We will need to update this before merging

Xiltyn replied:
Absolutely, but I needed that to test the gh-actions integration properly

mdjastrzebski commented on .github/workflows/docs-deploy.yml:
```diff
@@ -0,0 +1,40 @@
+name: Deploy Docs to GitHub Pages
```
Why two deploy scripts?

mdjastrzebski reviewed (approved):
Awesome work @Xiltyn! 💯

Xiltyn commented:
All updates have been committed and we're now pointing to main branch with our docs deployment pipeline.

@mdjastrzebski please note that I had to bump our main pipeline node version to 16.x since docusaurus project is using it. I can check if 14.x will work there but honestly I think that we should be using 16.x at this point anyway. Additionally, I added a step for running `yarn install` within the `./docusaurus` directory after the initial bob Installation and Build step. 

I also added .eslintignore file and created the ./docusaurus/.eslintrc file which extends our root .eslintrc and adds some docs specific rules since docusaurus is using some automagical imports ;) 

Please doubleck and if all is good we can merge tomorrow :)

thymikee commented on .github/workflows/main.yml:
```diff
@@ -17,12 +17,15 @@ jobs:
       - name: Install and build
         run: yarn install --force
 
+      - name: Install Docusaurus modules
+        run: cd ./docusaurus && yarn install --force
+
```
Is this even needed?

Xiltyn replied:
Feature wise - no. 

It's there because eslint throws errors due to to the fact that it cannot find imported modules in paths. E.g. `clsx`. 

So the options were at this moment:
1. Ignore whole `./docusaurus` folder and exclude it from linting checks altogether
2. Satisfy the linter. 

I went with the latter, since imo we should follow the chosen coding convention regardless on the submodule in the repo. 

If you have any other ideas how we could deal with it, I'm happy to look into it. Maybe we could add Docusaurus with Bob and migrate this integration to anothe sub-package, but it doesn't feel correct. After, it's not a part of the package itself and has nothing to do with the exported APIs. 

WDYT? @thymikee @mdjastrzebski

Xiltyn replied:
So, what are we doing with this guys?

thymikee replied:
Oh, that makes sense. Let's leave it then :)

thymikee commented on .github/workflows/main.yml:
```diff
@@ -17,12 +17,15 @@ jobs:
         uses: actions/setup-node@v3
         with:
-          node-version: 14.x
+          node-version: 16.x
```
I'd go with 18.x, if we only pick one, but 16 is good for now as well

Xiltyn replied:
I'm more than happy to go with 18.x, will update tomorrow :)

thymikee reviewed (approved):
Amazing!

mdjastrzebski commented:
Awesome work @Xiltyn The design is simply gorgeous!

On small nitpicking request, could you add trailing new lines in all files that miss them:
<img width="98" alt="image" src="https://user-images.githubusercontent.com/6368606/196797016-1f346ddd-9255-4ea5-b7ad-a36282638762.png">

Xiltyn commented:
> Awesome work @Xiltyn The design is simply gorgeous!
> 
> On small nitpicking request, could you add trailing new lines in all files that miss them: <img alt="image" width="98" src="https://user-images.githubusercontent.com/6368606/196797016-1f346ddd-9255-4ea5-b7ad-a36282638762.png">

Absolutely, will find a couple minutes tomorrow to make the last ammends so that we could actually merge :)

mdjastrzebski commented:
@Xiltyn merge when ready

Xiltyn commented:
All checks green and all review comments have been answered. 

**From now on, each new PR merged into main should start the documentation build.** 
@mdjastrzebski please keep an eye out if the actual behaviour deviates from the above

Cheer! 🚀

Xiltyn commented:
All checks green and all review comments have been answered. 

**From now on, each new PR merged into main should start the documentation build.** 
@mdjastrzebski please keep an eye out if the actual behaviour deviates from the above

Cheers! 🚀
